std-bits/table/src/main/java/org/enso/table/operations/AddGroupNumber.java

```diff
@@ -27,28 +16,51 @@ public static Storage<?> numberGroupsUnique(
     if (groupingColumns.length == 0) {
       throw new IllegalArgumentException("At least one grouping column is required.");
     }
+    var groupNumberRowVisitorFactory =
+        new GroupNumberRowVisitorFactory(start, step, Math.toIntExact(numRows));
+    GroupingOrderingVisitor.visit(
+        groupingColumns,
+        new Column[0],
+        new int[0],
+        problemAggregator,
+        groupNumberRowVisitorFactory,
+        numRows);
+    return new LongStorage(groupNumberRowVisitorFactory.numbers, IntegerType.INT_64);
+  }
 
-    var groupNumberIterator = new StepIterator(start, step);
+  private static class GroupNumberRowVisitorFactory implements RowVisitorFactory {
 
-    long[] numbers = new long[Math.toIntExact(numRows)];
+    private long current;
+    private final long step;
+    long[] numbers;
 
-    Storage<?>[] groupingStorages =
-        Arrays.stream(groupingColumns).map(Column::getStorage).toArray(Storage[]::new);
-    ColumnAggregatedProblemAggregator groupingProblemAggregator =
-        new ColumnAggregatedProblemAggregator(problemAggregator);
-    List<TextFoldingStrategy> textFoldingStrategy =
-        ConstantList.make(TextFoldingStrategy.unicodeNormalizedFold, groupingStorages.length);
-    Map<UnorderedMultiValueKey, Long> groupNumbers = new HashMap<>();
+    GroupNumberRowVisitorFactory(long start, long step, int size) {
+      this.current = start;
+      this.step = step;
+      numbers = new long[size];
+    }
 
-    for (int i = 0; i < numRows; i++) {
-      var key = new UnorderedMultiValueKey(groupingStorages, i, textFoldingStrategy);
-      key.checkAndReportFloatingEquality(
-          groupingProblemAggregator, columnIx -> groupingColumns[columnIx].getName());
-      var groupNumber = groupNumbers.computeIfAbsent(key, k -> groupNumberIterator.next());
-      numbers[i] = groupNumber;
+    @Override
+    public RowVisitor getNewRowVisitor() {
+      var nextGroupNumber = current;
+      current = Math.addExact(current, step);
+      return new GroupNumberRowVisitor(nextGroupNumber, numbers);
     }
 
-    return new LongStorage(numbers, IntegerType.INT_64);
+    private static class GroupNumberRowVisitor implements RowVisitor {
+      private final long groupNumber;
+      private final long[] numbers;
```

Comment:
It is not completely clear to me what this array represents. It can be figured out from the code but a comment would make it a bit easier to grasp.

Reply:
Rename it to returnStorage? returnRawStorage? returnValues?

Reply:
Sounds good.

Or maybe something like `storageForResult`?